### PR TITLE
fix(grid3d): Memory usage reduced.

### PR DESCRIPTION
**Pull request by LeonidPolukhin**
1) Memory usage is significantly reduced in Grid3d webworker.
2) If there is not enough memory to hold WebGL meshes size of output data arrays is reduced by 10% until success.
3) In addition to JSONs , Grid3d can load input data from raw binary files.
4) Common type definitions are extracted into separate file.

### CODE REVIEW

File: typescript/packages/subsurface-viewer/src/layers/grid3d/grid3dLayer.ts [hunk 1]
```diff
@@ -85,8 +85,17 @@ async function loadData<T extends TTypedArray>(
         return new type(data);
     }
     if (typeof data === "string") {
-        const stringData = await load(data as string, JSONLoader);
-        return new type(stringData);
+        const extension = data.split(".").pop()?.toLowerCase();
```
Comment (w1nklr) [inline typescript/packages/subsurface-viewer/src/layers/grid3d/grid3dLayer.ts:88]: Add comments for expected data (json file, or ??)

File: typescript/packages/subsurface-viewer/src/layers/grid3d/webworker.ts [hunk 2]
```diff
@@ -839,6 +833,95 @@ export function makeFullMesh(e: { data: WebWorkerParams }) {
...
+        properties: Float32Array;
+        lineIndices: Uint32Array;
+    }
+
+    const getPrimitiveCounts = (polys: Uint32Array): IPrimitiveCounts => {
+        let triangles = 0;
+        let lineSegments = 0;
+        let i = 0;
+        while (i < polys.length) {
+            triangles += polys[i] - 2;
```
Comment (w1nklr) [inline typescript/packages/subsurface-viewer/src/layers/grid3d/webworker.ts:853]: Why -2 ?

File: typescript/packages/subsurface-viewer/src/layers/grid3d/webworker.ts [hunk 3]
```diff
@@ -839,6 +833,95 @@ export function makeFullMesh(e: { data: WebWorkerParams }) {
...
+            attributes: {
+                positions: { value: new Float32Array(), size: 3 },
+                indices: { value: new Uint32Array(), size: 1 },
+            },
+            vertexCount: 0,
+        };
+        return [
+            mesh,
+            mesh_lines,
+            [propertyValueRangeMin, propertyValueRangeMax],
```
Comment (w1nklr) [inline typescript/packages/subsurface-viewer/src/layers/grid3d/webworker.ts:921]: Why the range here ?

File: typescript/packages/subsurface-viewer/src/layers/grid3d/webworker.ts [hunk 4]
```diff
@@ -839,6 +833,95 @@ export function makeFullMesh(e: { data: WebWorkerParams }) {
...
+                properties,
+                lineIndices,
+            };
+        } catch (error) {
+            console.log(error);
+            return null;
+        }
+    };
+
+    const createMeshArrays = (
```
Comment (w1nklr) [inline typescript/packages/subsurface-viewer/src/layers/grid3d/webworker.ts:878]: Small doc

File: typescript/packages/subsurface-viewer/src/layers/grid3d/webworker.ts (hunk 2 above)
Comment (LeonidPolukhin) [reply typescript/packages/subsurface-viewer/src/layers/grid3d/webworker.ts:853]: 3 points - 1 triangle, 4 points - 2 triangles.. etc.

File: typescript/packages/subsurface-viewer/src/layers/grid3d/webworker.ts (hunk 3 above)
Comment (LeonidPolukhin) [reply typescript/packages/subsurface-viewer/src/layers/grid3d/webworker.ts:921]: Properties are per face in the input array.  As the webworker assignes a property value for every vertex it also can calc its range and return to the main thread

File: typescript/packages/subsurface-viewer/src/layers/grid3d/grid3dLayer.ts (hunk 1 above)
Comment (LeonidPolukhin) [reply typescript/packages/subsurface-viewer/src/layers/grid3d/grid3dLayer.ts:88]: Comment added

File: typescript/packages/subsurface-viewer/src/layers/grid3d/webworker.ts (hunk 4 above)
Comment (LeonidPolukhin) [reply typescript/packages/subsurface-viewer/src/layers/grid3d/webworker.ts:895]: Added

File: typescript/packages/subsurface-viewer/src/layers/grid3d/webworker.ts (hunk 2 above)
Comment (w1nklr) [reply typescript/packages/subsurface-viewer/src/layers/grid3d/webworker.ts:858]: OK. The polys are defined as triangle strip.
It's worth a comment...

File: typescript/packages/subsurface-viewer/src/layers/grid3d/webworker.ts (hunk 3 above)
Comment (w1nklr) [reply typescript/packages/subsurface-viewer/src/layers/grid3d/webworker.ts:942]: These are never used. If you rename them, it still does work :/

Comment (LeonidPolukhin) [reply typescript/packages/subsurface-viewer/src/layers/grid3d/webworker.ts:942]: They are expected to be returned from the webworker

File: typescript/packages/subsurface-viewer/src/layers/grid3d/webworker.ts (hunk 2 above)
Comment (LeonidPolukhin) [reply typescript/packages/subsurface-viewer/src/layers/grid3d/webworker.ts:861]: Comments added, docs updated

File: typescript/packages/subsurface-viewer/src/layers/grid3d/webworker.ts (hunk 3 above)
Comment (w1nklr) [reply typescript/packages/subsurface-viewer/src/layers/grid3d/webworker.ts:945]: OK, I miss-read the PR.

**Review by w1nklr — APPROVED** [review]

**Pull request merged**